File: Niveau 3/Niveau 3.9/palyndrome.hs
--------------------------------------------------------------------------------
-- Name:        Plus long palindrome
-- Purpose:     France-IOI niveau 3.9
--
-- Author:      Sébastien
--
-- Created:     09/10/2022
--------------------------------------------------------------------------------

main :: IO()
main = do  
    texte <- getLine
    print $ maximum (map (chercherPalindromeMax texte) [0..length texte -1])
    
chercherPalindromeMax :: String -> Int -> Int
chercherPalindromeMax [] _ = 0
chercherPalindromeMax texte position = let gauche = reverse $ take position texte
                                           droite = drop position texte
                                           droite' = drop (position+1) texte
        in max (chercherPalindrome gauche droite) (1 + chercherPalindrome gauche droite')

chercherPalindrome :: String -> String -> Int
chercherPalindrome [] _ = 0
chercherPalindrome _ [] = 0
chercherPalindrome (g:gs) (d:ds) = if g == d then 2 + chercherPalindrome gs ds else 0

File: Niveau 3/Niveau 3.9/palyndrome.py
#-------------------------------------------------------------------------------
# Name:        Plus long palindrome
# Purpose:     France-IOI niveau 3.9
#
# Author:      Sébastien
#
# Created:     09/10/2022
#-------------------------------------------------------------------------------

def chercherPalindrome(texte, gauche, droite):
    while gauche >= 0 and droite < len(texte) and texte[gauche] == texte[droite]:
        gauche -= 1
        droite += 1
##    print("[" + texte[gauche+1:droite] + "]", end=' ')
    return droite - gauche - 1

def chercherPalindromeMaximum(texte):
    maxi = 1
    for i in range(len(texte)):
##        print(i, texte[i], ":", end=' ')
        if i + maxi//2 >= len(texte): break
        maxi = max(maxi, chercherPalindrome(texte, i-1, i+1))
        maxi = max(maxi, chercherPalindrome(texte, i, i+1))
##        print()
    return maxi

if __name__ == '__main__':
    texte = input()
    print(chercherPalindromeMaximum(texte))
